Rule 2ee8b8 ("Visible label is part of accessible name"): introducing a new "label in name algorithm". (#2075)
auto-wcag/auto-wcag: _rules/visible-label-in-accessible-name-2ee8b8.md

- The element has a [semantic role][] that is a [widget][widget role] that [supports name from content][]; and
- The element has [visible text content][]; and

> - The element has an `aria-label` or `aria-labelledby` attribute; and

Reviewer: **Chore:** there seem to be no Applicable examples with `aria-labelledby`. OTOH, this was also not the case previously, so we can keep that for a later PR…

If the target element contains an [image of text](https://www.w3.org/TR/WCAG22/#label-in-name), it may pass this rule but fail [2.5.3 Label in Name][understand253] (because the accessible name should then match the text inside the image, which is not taken into account by this rule). So further testing is needed. This case might be handled by a different rule in the future.

### Assumptions

This rule assumes that the [visible inner text][] is equal to the [label as defined by WCAG][(https://www.w3.org/wai/wcag22/understanding/label-in-name#dfn-label), even though "label" is not precisely defined at this point in history.

This rule assumes that the visible label isn't rearranged with CSS so that it appears to the user in a different order than it appears in the DOM.

This rule assumes that the visible label doesn't use CSS to add whitespace where none exists in the DOM.

> This rule - specifically, the [label in name algorithm][] that this rule relies on - assumes that content within parentheses can be ignored.  ("Parentheses" are also known as "round brackets".)  This is important becuase the algorithm's treatment of parentheses is to remove them and all characters within them. This assumption is almost always true in English.  Exceptions include links with names such "Dune (1984 film)" and "Dune (2021 film)". This assumption is known to be often false in languages other than English, such as German (where parentheses indicate dual states) and Arabic (where parentheses are often used as quotation marks). Violations of this assumption will, in real-world scenarios, more often result in a false negative for this rule rather than a false positive.

Reviewer: ```suggestion
This rule — specifically, the [label in name algorithm][] that this rule relies on — assumes that content within parentheses can be ignored ("Parentheses" are also known as "round brackets"). This is important because the algorithm's treatment of parentheses is to remove them and all characters within them. This assumption is almost always true in English.  Exceptions include links with names such "Dune (1984 film)" and "Dune (2021 film)". This assumption is known to be often false in languages other than English, such as German (where parentheses indicate dual states) and Arabic (where parentheses are often used as quotation marks). Violations of this assumption will, in real-world scenarios, more often result in a false negative for this rule rather than a false positive.
```

Author: I manually applied this suggestion in commit 4002eb2c84519b17497121017850a2be918b79ea

- The element has a [semantic role][] that is a [widget][widget role] that [supports name from content][]; and
- The element has [visible text content][]; and
- The element has an `aria-label` or `aria-labelledby` attribute.
- The element does not contain any [rendered image resources][rendered image resource].

> - If there are any words that appear in both the element's accessible name and its visible label, then for each such pair, the same spelling and hyphenation is used in both places.

Reviewer 2: I'm not clear why this (line 41) is under Applicability. Maybe an Inapplicable Example would help.

Author: It used to be in the assumptions, then I moved it into the applicability because that was Giacomo's idea in the meeting on 2026-02-05.  The logic was, as the [minutes ](https://www.w3.org/2026/02/05-act-r-minutes.html) put it, that it would "remove from applicability [any case] that has no consensus".

At any rate, I just added an inapplicable example in commit 903ee9db28c7c1ed7492f3f45c1f2e3d908f0257.

- The element has a [semantic role][] that is a [widget][widget role] that [supports name from content][]; and
- The element has [visible text content][]; and
- The element has an `aria-label` or `aria-labelledby` attribute.
- If there are any words that appear in both the element's accessible name and its visible label, then for each such pair, the same spelling and hyphenation is used in both places.

> - Both the element's accessible name and its visible label do not contain any abbreviations.

Reviewer: ```suggestion
- The element has an `aria-label` or `aria-labelledby` attribute; and
- Neither the element's accessible name nor its visible label contains any abbreviations; and
- Every word that appear in both the element's accessible name and its visible label uses the same spelling and hyphenation in both places.
```
Reformated it as a conditions list, and rephrased conditions to flow more naturally (I think).

## Expectation

> For the target element, the [visible inner text][] is contained within the [accessible name][] according to the [label in name algorithm][].

Reviewer: ```suggestion
For each target element, the [visible inner text][] is contained within the [accessible name][] according to the [label in name algorithm][].
```
ACT rules use the "For each [target]" phrasing in Expectation.

## Background

This rule applies to elements with a [widget role][] that [support name from content][supports name from content]. This includes the following: `button`, `checkbox`, `gridcell`, `link`, `menuitem`, `menuitemcheckbox`, `menuitemradio`, `option`, `radio`, `searchbox`, `switch`, `tab`, `treeitem`.

The understanding document of [2.5.3 Label in Name][understand253] use the term "symbolic text characters" to refer to a type of [non-text content][] that uses text characters as symbols, such as using "x" to mean "close". This rule considers them as "characters expressing non-text content". Unicode emojis are another example of characters expressing non-text content, although these are not "symbolic text characters".

> Regarding images of text: if the target element contains an [image of text](https://www.w3.org/WAI/WCAG22/Understanding/label-in-name.html#dfn-image-of-text), it may pass this rule but fail [2.5.3 Label in Name][understand253].  So further testing is needed.  This case might be handled by a different rule in the future.

Reviewer: ```suggestion
If the target element contains an [image of text](https://www.w3.org/WAI/WCAG22/Understanding/label-in-name.html#dfn-image-of-text), it may pass this rule but fail [2.5.3 Label in Name][understand253] (because the accessible name should then match the text inside the image, which is not taken into account by this rule). So further testing is needed. This case might be handled by a different rule in the future.
```
Also, the link here should be to the SC in WCAG, not to the Understanding document.

### Assumptions

> This rule assumes that the [visible inner text][] is equal to the [label][https://www.w3.org/wai/wcag21/understanding/label-in-name#dfn-label], even though "label" is not precisely defined at this point in history.

Reviewer: ```suggestion
This rule assumes that the [visible inner text][] is equal to the [label as defined by WCAG][(ttps://www.w3.org/wai/wcag21/understanding/label-in-name#dfn-label), even though "label" is not precisely defined at this point in history.
```

This rule assumes that the visible label isn't rearranged with CSS so that it appears to the user in a different order than it appears in the DOM.

This rule assumes that the visible label doesn't use CSS to add whitespace where none exists in the DOM.

> This rule - specifically, the [label in name algorithm][] that this rule relies on - assumes that the algorithm's treatment of parentheses is appropriate in the given human language. "Parentheses" are also known as "round brackets". The algorithm's treatment of parentheses is to remove them and all characters within them. This assumption can be reworded as: content within parentheses can be ignored. This assumption is almost always true in English. It is known to be often false in other languages, such as German (where parentheses indicate dual states) and Arabic (where parentheses are often used as quotation marks). Violations of this assumption will, in real-world scenarios, more often result in a false negative for this rule rather than a false positive.

Reviewer: **Issue:** I am a bit uncomfortable in stating an assumption and immediately saying that it is often not met 😓 
I think the bottom line is OK as it is indeed, "only" false negatives. But I think it should be rephrased and maybe moved to Background (like the "image of text" discussion).
Something like "In languages where parenthesis are used for non-incidental [or whatever the correct grammatical term] content, this rule may pass while 2.5.3 fails because a parenthesis is present on one side and its content matter. Examples of such languages include Arabic and German [putting the languages in alphabetical order]" (with a bit more meat to that text, but essentially what is in the current one).

Author: I'm not sure I completely agree.  I am very comfortable stating an assumption and immediately saying that it is often not met ... if that's true and helpful.  I have spent enough hours trying to decipher specs - WCAG, ACT, you name it - wishing they would be more straightforward and not try to hide their warts.  I think that WCAG has a serious problem in this area.  ACT doesn't AFAICS, but I am loathe to move the needle even the tiniest amount in that direction.  Your thoughts?

Reviewer: I'm certainly not advocating to remove it totally, but moving it to background like the "image of text" note. That note could as well be phrased as "this rule assumes that the page has not image of text".
I'd rather keep assumptions for rare cases.

Not blocking for me however.

Author: I see this as an argument to move the "image of text" note to the "Assumptions" section, not to move the parentheses note to the "Background" section.  :)   

I had a conversation with someone you know at CSUN.  He also saw assumption violations as rare things.  I guess I don't.  I just looked at my database and found 127 assumption violations for alfa's target size rules.  Customers dispute those due to the "equivalent" exception of the SC == assumption of the alfa rule.  We've had the target size rules for exactly two years.   127/(52*2) = 1.2 assumption violations per week.    Whether that meets the definition of "rare" depends on one's perspective.  It's small compared to other things.   But still happens regularly (= 1.2 times per week).

At any rate, if it's not a blocker for you, I'll leave it as is - and thank you for the discussion.  If someone else weighs in, I'll take it from there.

## Expectation

For the target element, the [visible inner text][] is contained within the [accessible name][] according to the [label in name algorithm][].

## Assumptions

This rule assumes that the [visible inner text][] is equal to the [label][https://www.w3.org/WAI/WCAG21/Understanding/label-in-name#dfn-label], even though "label" is not precisely defined at this point in history.

This rule assumes that the visible label isn't rearranged with CSS so that it appears to the user in a different order than it appears in the DOM.

This rule assumes that the visible label doesn't use CSS to add whitespace where none exists in the DOM.

> This rule assumes that for any word which appears in both the accessible name and the visible label, the same spelling and hyphenation is used in both places.  For example: if "non-negative" is used in the accessible name and "nonnegative" is used in the visible label, that would violate this assumption.  Or if "color" is used in the accessible name and "colour" is used in the visible label, that would also violate this assumption.

Reviewer 3: ```suggestion
This rule assumes that for any word which appears in both the accessible name and the visible label, the same spelling and hyphenation is used in both places.  For example, if "non-negative" is used in the accessible name and "nonnegative" is used in the visible label, that would violate this assumption.  Similarly, if "color" is used in the accessible name and "colour" is used in the visible label, that would also violate this assumption.
```

### Assumptions

This rule assumes that the [visible inner text][] is equal to the [label][https://www.w3.org/wai/wcag21/understanding/label-in-name#dfn-label], even though "label" is not precisely defined at this point in history.

This rule assumes that the visible label isn't rearranged with CSS so that it appears to the user in a different order than it appears in the DOM.

This rule assumes that the visible label doesn't use CSS to add whitespace where none exists in the DOM.

> This rule assumes that for any word which appears in both the accessible name and the visible label, the same spelling and hyphenation is used in both places. For example: if "non-negative" is used in the accessible name and "nonnegative" is used in the visible label, that would violate this assumption. Similarly, if "color" is used in the accessible name and "colour" is used in the visible label, that would also violate this assumption.

Reviewer 4: I believe the presence or absence of a hyphen is exempt under the 'punctuation optional' wording in the "Punctuation and capitalization" section of the Understanding document. Speech recognition users would rarely say "non hyphen negative" or "non dash negative". Most would just say "nonnegative"; my personal experience with speech recognition is that it would generative a match in either case.

Author: I agree - and I think that's why it's appropriate here, under "Assumptions".  ACT assumptions is the home for things that, if broken, can result in a case that fails the rule and passes the SC, which is what this is, am I right?

Reviewer 4: I'm saying I think your assumption is too contained in this case. It's relatively easy to ignore hyphens, so I think a better assumption would be that hyphens, as punctuation, are ignored. I agree that misspellings are more problematic. Short of a dictionary of allowed spelling variances, I think that you starting assumption for spelling is supportable.

> This rule - specifically, the [label in name algorithm][] that this rule relies on - assumes that the algorithm's treatment of parentheses is appropriate in the given human language.  "Parentheses" are also known as "round brackets".  The algorithm's treatment of parentheses is to remove them and all characters within them.  This assumption can be reworded as: content within parentheses can be ignored.  This assumption is almost always true in English.  It is known to be often false in other languages, such as German (where parentheses indicate dual states) and Arabic (where parentheses are often used as quotation marks).  Violations of this assumption will, in real-world scenarios, more often result in a false negative for this rule rather than a false positive.

Reviewer 4: However, if all the information in parentheses is stripped out of the test, wouldn't a resulting match for the remaining string still be valid, regardless of language?
i.e., If I strip out the parenthetical material in "This is my label (XXXX)" from both the label and name strings, shouldn't the resulting name still contain the resulting string of the label?

Author: That sounds correct to me.  Does that cause any problems?

Reviewer 5: I would suggest the algorithm has a first step for [text normalization](https://en.wikipedia.org/wiki/Text_normalization#:~:text=Text%20normalization%20is%20the%20process,operations%20are%20performed%20on%20it.) before any comparison.

Reviewer 4: > That sounds correct to me. Does that cause any problems?

I agree that for variable considerations like Label in Name, it makes sense to bias tests towards false negatives (that is, to err on the side of flagging fewer potential issues ).

Reviewer 6: My suggestion to this problem is:

1. Don't remove content in parentheses, don't remove punctuation. Instead allow some optional content to be defined by the user in a design system. In the case of required fields, this information is should be presented to AT through a required attribute. By allowing the design system define what this looks like, we can tell testers to expect one format for visually displaying the field is required. For example, they can define `/^.(\(required\))$/` or `/^. - required$/` as the required field format, that text can be removed before performing the test, but is can also be used to test that the label passes a future test for [SC 3.2.4](https://w3.org/TR/WCAG21#consistent-identification). This also solves for the i18n issue as we push the regionalization aspect back to the owners.
1. Despite what I said in my last comment, I think we should open a relationship with the W3C internationalization working group, they may have resources that can help to validate if any of our rules present problems for specific locales.
1. We should work with WCAG 3.x writers to get 2.5.3 updated to go beyond label in name, and instead the label should be in the name and state. Until that happens, I don't think the discussion around required fields has a basis in WCAG, even if it is the right thing to do.

Reviewer 6: Edit: I just re-read the understanding section of 2.5.3 and it seems more clear where these decisions came from, but I still think establishing a site-wide pattern (through a design system) is the best way to encapsulate the ambiguity. Let the design system define a format for labels, for example:

```js
/^(?<label>.*?)(?<description>\(.*\))?(?<required> - required)?(?<punctuation>:)$/
```

This pattern would allow the following labels:

* Date (DD/MM/YYYY) - required:
  * label: "Date"
  * description: "DD/MM/YYYY
  * required: `true`
  * Ignored punctuation: ":"
* First name - required:
  * label: "First name"
  * description: `NULL`
  * required: `true`
  * Ignored punctuation: ":"
* Middle name:
  * label: "Middle name"
  * description: `NULL`
  * required: `false`
  * Ignored punctuation: ":"

It would only allow the following labels if the string is a substring of the accessible name
* Date (DD/MM/YYYY) - required
* First name (required)

At a later date, we can use this same definition to write a partial test for [SC 3.2.4](https://w3.org/TR/WCAG21#consistent-identification) to insist that the labels follow the consistent identification as defined in the design system.

It also allows us to create a partial test for [SC 1.3.1](https://w3.org/TR/WCAG21#info-and-relationships) to validate that the required attribute is represented in the accessibility tree.

Author: As much as I see promise in the ideas that you've written, I don't see how I can act on them in this PR.
1. The idea of "establishing a site-wide pattern (through a design system)" is something I don't know how to do in an ACT rule.
1. &lt;label&gt; cases aren't applicable under this rule.  This is not clear in the rule, nor in what I said on Thursday, but I've since read it again, and I'm quite sure that &lt;label&gt; cases are not applicable, because they don't support "name from content" (even though they are deceptively similar).  I think we all - myself included - have gotten side-tracked around these &lt;label&gt; cases eg. a "First name (required)" text field.  I should have seen that earlier, and saved us some time.  Note the example in this PR which has brackets - currently Passed Example 14, "Search by date (YYYY-MM-DD)" - is a plain button (not a text field).  So it's rather contrived. 
1. In so far as your ideas fix false negatives: I think that we should not worry about false negatives at this time and place.  In this I agree with mbgower's comment above that "it makes sense to bias tests towards false negatives".  Indeed, that is a pattern that I've seen in all ACT rule-writing.  This pattern seems to be documented at https://www.w3.org/WAI/standards-guidelines/act/rules/about/ in the sections "ACT Rules are Partial Checks" and "ACT Rules Check for Failures".

#### Passed Example 1

This link has [visible inner text][] that is equal to the [accessible name][].

```html
<a href="https://act-rules.github.io/" aria-label="ACT rules">ACT rules</a>
```

#### Passed Example 2

This link has [visible inner text][] that, ignoring whitespace, is equal to the [accessible name][].

```html
<a href="https://act-rules.github.io/" aria-label="  ACT   rules  ">ACT rules</a>
```

#### Passed Example 3

This link has [visible inner text][] that, ignoring case, is equal to the [accessible name][].

```html
<a href="https://act-rules.github.io/" aria-label="act Rules">ACT rules</a>
```

#### Passed Example 4

This button has [visible inner text][] that is contained within the [accessible name][] according to the [label in name algorithm][].

```html
<button aria-label="Next Page in the list">Next Page</button>
```

#### Passed Example 5

> This button has [visible inner text][] that does not need to be contained within the [accessible name][], because the "x" text node is [non-text content][]. Note: this would need to meet SC 1.1.1 Non text content.

Reviewer 2: In the algorithm, 

> For each character that either a) represents non-text content, or b) isn't a letter or a digit: replace that character with a space character.

If "X" is replaced with a space char, why would this example pass?

#### Passed Example 1

This link has [visible inner text][] that is equal to the [accessible name][].

```html
<a href="https://act-rules.github.io/" aria-label="ACT rules">ACT rules</a>
```

#### Passed Example 2

This link has [visible inner text][] that, ignoring whitespace, is equal to the [accessible name][].

```html
<a href="https://act-rules.github.io/" aria-label="  ACT   rules  ">ACT rules</a>
```

#### Passed Example 3

This link has [visible inner text][] that, ignoring case, is equal to the [accessible name][].

```html
<a href="https://act-rules.github.io/" aria-label="act Rules">ACT rules</a>
```

#### Passed Example 4

This button has [visible inner text][] that is contained within the [accessible name][] according to the [label in name algorithm][].

```html
<button aria-label="Next Page in the list">Next Page</button>
```

#### Passed Example 5

> This button has [visible inner text][] that does not need to be contained within the [accessible name][], because the "x" text node is [non-text content][]. Note: this would need to meet SC 1.1.1 Non text content.

Reviewer: ```suggestion
This button has [visible inner text][] that does not need to be contained within the [accessible name][], because the "X" text node is [non-text content][]; the [label in name algorithm][] therefore ignores it. Note: this would need to meet SC 1.1.1 Non text content.
```

```html
<button aria-label="anything">X</button>
```

#### Passed Example 6

> This `button` element has the text "search" rendered as a magnifying glass icon by the font. Because the text is rendered as [non-text content][], the text does not need to be contained within the [accessible name][].

Reviewer: ```suggestion
This `button` element has the text "search" rendered as a magnifying glass icon by the font. Because the text is rendered as [non-text content][], the text does not need to be contained within the [accessible name][]; and the [label in name algorithm][] ignores it.
```

<a aria-label="Download specification" href="#">Download <span class="visually-hidden">gizmo</span> specification</a>
```

#### Passed Example 11

This example shows that the [label in name algorithm][] handles many kinds of whitespace.

```html
<a aria-label="compose email" href="#">compose &nbsp;&nbsp;<br> email</a>
```

#### Passed Example 12

This example passes the rule because "YYYY-MM-DD" is in brackets.  Text in brackets is removed by the [label in name algorithm][], because its not normally spoken by speech-input users.  

```html
<button aria-label="Search by date">Search by date (YYYY-MM-DD)</button>
```

#### Passed Example 13

The passes for two reasons: 1) because the ellipsis ("…") is [non-text content][], and 2) because the ellipsis is neither a letter nor a digit and so is filtered out by the [label in name algorithm][].

```html
<button aria-label="Next">Next…</button>
```

#### Passed Example 14

This passes because the [label in name algorithm][] effectively ignores all punctuation and emoji, in both the visible inner text and the accessible name, as long as they don't break up words.

```html
<button aria-label="💡 Submit 💡">>>> ** Submit ** <<<</button>
```

> #### Passed Example 15
> 
> The "X" is non-text content.
> 
> ```html
> <button aria-label="Close">X</button>
> ```

Reviewer: ```suggestion
```
(this is the same as Passed Example 5)

Author: Thanks, I just pushed a commit for this.

This button has [visible inner text][] that is only partially contained within the [accessible name][].

```html
<button aria-label="the full">The full label</button>
```

#### Failed Example 3 

This button has [visible inner text][] that is fully contained within the [accessible name][] when viewed as a character-by-character substring.  But that does not satisfy our [label in name algorithm][], which works on full words.  So this fails the rule.

```html
<a href="#" aria-label="Discover Italy">Discover It</a>
```

#### Failed Example 4 

This link's [accessible name][] contains two tokens (according to the[label in name algorithm][]) and the [visible inner text][] contains one token.  So it fails the rule.

```html
<a aria-label="just ice" href="#">justice</a>
```

#### Failed Example 5 

This link has an [accessible name][] which contains a hyphen.  The [label in name algorithm][] breaks up words on hyphens.  So it turns "non-standard" into two tokens: "non" and "standard".  So this fails the rule.

```html
<a href="#" aria-label="non-standard">nonstandard</a>
```

#### Failed Example 6 

The rule has no special handling for acronyms or initialisms.

```html

> <a aria-label="WAVE" href="#">W A V E</a>

Reviewer: We should take a name that isn't also an Accessibility vendor.
Maybe simply "WCAG"?

Author: Done.

<button aria-label="Find">search</button>
```

#### Passed Example 7 

This button has [visible inner text][] that, according to the [label in name algorithm][], is contained within the [accessible name][].  This example shows why the [label in name algorithm][] uses the [visible inner text][] and not the [visible text content][]: the <p> tags insert whitespace into the result in the former but not the latter.

```html
<button aria-label="Hello world"><p>Hello</p><p>world</p></button>
```

#### Passed Example 8

> Similar to the previous example.

Reviewer: Referencing "previous example" is a bit risky since nothing guarantees they will stay in the same order next time we rewrite the rule.

Author: I agree there is a risk.  In this case I think that the risk is worth it.   With "previous", I claim that the readability is improved now.  If I remove "previous", I can't think of a way to rewrite it that doesn't make it significantly less readable.

Reviewer 2: Maybe include additional description, like: 

```suggestion
Similar to the previous example. The [label in name algorithm][] inserts whitespace.
```

```html
<a href="#" aria-label="Some article by John Doe"><h6>Some article</h6><p>by John Doe</p></a>
```

#### Passed Example 9

> The [visible inner text][] of this link is "ACT" (with no spaces) because of the explicit styles of "display: inline" on the &lt;p&gt; elements and the absence of whitespace between the &lt;p&gt; elements.  The cases of "display: inline" and "display: block" are handled by the definition of [visible inner text of an element][].  This example shows that the definition agrees with the visual rendering done by the browser.

Reviewer: ```suggestion
The [visible inner text][] of this link is "ACT" (with no spaces) because of the explicit styles of `display: inline` on the `p` elements and the absence of whitespace between the `p` elements.  The cases of `display: inline` and `display: block` are handled by the definition of [visible inner text of an element][].  This example shows that the definition agrees with the visual rendering done by the browser.
```

Author: Done in 2dcd941d3a853e5dcee6acfc0332ac9de1b72515.

```

#### Passed Example 13

This example shows that the [label in name algorithm][] handles many kinds of whitespace.

```html
<a aria-label="compose email" href="#">compose &nbsp;&nbsp;<br> email</a>
```

#### Passed Example 14

This example passes the rule because "YYYY-MM-DD" is in brackets.  Text in brackets is removed by the [label in name algorithm][], because its not normally spoken by speech-input users.  

```html
<button aria-label="Search by date">Search by date (YYYY-MM-DD)</button>
```

#### Passed Example 15

The passes for two reasons: 1) because the ellipsis ("…") is [non-text content][], and 2) because the ellipsis is neither a letter nor a digit and so is filtered out by the [label in name algorithm][].

```html
<button aria-label="Next">Next…</button>
```

#### Passed Example 16

This passes because the [label in name algorithm][] effectively ignores all punctuation and emoji, in both the visible inner text and the accessible name, as long as they don't break up words.

```html
<button aria-label="💡 Submit 💡">&gt;&gt;&gt; ** Submit ** &lt;&lt;&lt;</button>
```

> 

Reviewer 7: I think adding the "First Name:" pass example from the Understanding document would be useful (since it's very common).

Author: It's a strong example for the SC, but I can't see how to adapt it to this ACT rule.  That example (`<label for="firstname">First Name:</label> <input id="firstname" type="text" name="firstname">`) doesn't meet the applicability of this ACT rule because `<input>` doesn't support ["name from content"](https://www.w3.org/TR/wai-aria-1.2/#namefromcontent).  For us to add an example which uses that visible label "First Name:" on an element which does support "name from content" (mostly `<a>` and `<button>`) seems unrealistic to me.  Suggestions welcome on how to reconcile this.

```html
<a aria-label="Download specification" href="#"
	><span>Download</span><span id="space"> </span><span>specification</span></a
>
```

#### Passed Example 12

This example shows that the [visible inner text][] isn't always the same as the [innerText][https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute]. The visible inner text is "Download specification". The innerText is 'Download \ngizmo\nspecification'. This rule uses the visible inner text - not innerText.

```html
<style>
	.visually-hidden {
		/* Source: https://www.tpgi.com/the-anatomy-of-visually-hidden/ */
		clip-path: inset(50%);
		height: 1px;
		overflow: hidden;
		position: absolute;
		white-space: nowrap;
		width: 1px;
	}
</style>
<a aria-label="Download specification" href="#">Download <span class="visually-hidden">gizmo</span> specification</a>
```

#### Passed Example 13

This example shows that the [label in name algorithm][] handles many kinds of whitespace.

```html
<a aria-label="compose email" href="#">
	compose &nbsp;&nbsp;<br />
	email
</a>
```

#### Passed Example 14

> This example passes the rule because "YYYY-MM-DD" is in brackets. Text in brackets is removed by the [label in name algorithm][], because it is not normally spoken by speech-input users.

Reviewer 2: Since square or curly brackets aren't mentioned in the algorithm,

```suggestion
This example passes the rule because "YYYY-MM-DD" is in round brackets. Text in round brackets is removed by the [label in name algorithm][], because it is not normally spoken by speech-input users.
```

```html
<button aria-label="11 times 3 equals 33">11×3=33</button>
```

#### Failed Example 8

This button's accessible name contains the same tokens that are in the visible label. But they aren't in the same order, so it fails the sublist check part of the [label in name algorithm][], and so it fails the rule.

```html
<button aria-label="how are you"><span>you</span><span>how</span><span>are</span></button>
```

#### Failed Example 9

This button's accessible name contains the word "the" in the middle of it, which causes the sublist check of the [label in name algorithm][] (in particular: the "consecutive" requirement of that check) to fail. So it fails the rule.

```html
<button aria-label="Download the specification">Download specification</button>
```

#### Failed Example 10

This link's accessible name contains the same digits that are in the visible label, and in the same order. But they have different spaces and punctuation between them, so they are considered separate tokens. So this fails the rule.

```html
<a aria-label="1 2 3. 4 5 6. 7 8 9 0" href="tel:1234567890">123.456.7890</a>
```

#### Failed Example 11

> This rule has no special handling which tries to guess when number have the same semantic meaning. It operates on tokens only.

Reviewer 2: ```suggestion
This rule has no special handling which tries to guess when numbers have the same semantic meaning. It operates on tokens only.
```

```html
<a href="#2021" aria-label="20 21">2021</a>
```

#### Failed Example 12

> Similar to the previous example.

Reviewer 2: ```suggestion
Similar to the previous example. This rule has no special handling which tries to guess when numbers have the same semantic meaning. It operates on tokens only.
```

```html
<style>
	.visually-hidden {
		/* Source: https://www.tpgi.com/the-anatomy-of-visually-hidden/ */
		clip-path: inset(50%);
		height: 1px;
		overflow: hidden;
		position: absolute;
		white-space: nowrap;
		width: 1px;
	}
</style>
<a aria-label="Download specification" href="#">Download <span class="visually-hidden">gizmo</span> specification</a>
```

#### Passed Example 13

This example shows that the [label in name algorithm][] handles many kinds of whitespace.

```html
<a aria-label="compose email" href="#"
	>compose &nbsp;&nbsp;<br />
	email</a
>
```

#### Passed Example 14

This example passes the rule because "YYYY-MM-DD" is in brackets. Text in brackets is removed by the [label in name algorithm][], because its not normally spoken by speech-input users.

```html
<button aria-label="Search by date">Search by date (YYYY-MM-DD)</button>
```

#### Passed Example 15

> The passes for two reasons: 1) because the ellipsis ("…") is [non-text content][], and 2) because the ellipsis is neither a letter nor a digit and so is filtered out by the [label in name algorithm][].

Reviewer 4: ```suggestion
This passes for two reasons: 1) because the ellipsis ("…") is [non-text content][], and 2) because the ellipsis is neither a letter nor a digit and so is filtered out by the [label in name algorithm][].
```

```html
<button aria-label="💡 Submit 💡">&gt;&gt;&gt; ** Submit ** &lt;&lt;&lt;</button>
```

### Failed

#### Failed Example 1

This link has [visible inner text][] that is very different from the [accessible name][].

```html
<a href="https://act-rules.github.io/" aria-label="WCAG">ACT rules</a>
```

#### Failed Example 2

This button has [visible inner text][] that is only partially contained within the [accessible name][].

```html
<button aria-label="the full">The full label</button>
```

#### Failed Example 3

This button has [visible inner text][] that is fully contained within the [accessible name][] when viewed as a character-by-character substring. But that does not satisfy our [label in name algorithm][], which works on full words. So this fails the rule.

```html
<a href="#" aria-label="Discover Italy">Discover It</a>
```

#### Failed Example 4

This link's [accessible name][] contains two tokens (according to the[label in name algorithm][]) and the [visible inner text][] contains one token. So it fails the rule.

```html
<a aria-label="just ice" href="#">justice</a>
```

#### Failed Example 5

> This link has an [accessible name][] which contains a hyphen. The [label in name algorithm][] breaks up words on hyphens. So it turns "non-standard" into two tokens: "non" and "standard". So this fails the rule.

Reviewer 4: > The [label in name algorithm][] breaks up words on hyphens

That seems somewhat problematic, especially when considered from a speech input perspective, where aurally there is no meaningful difference between "non standard", "non-standard", and "nonstandard".

Author: I agree, and it's unclear to me what the ideal fix for this would be, but I think this is a problem which - as I've said for other problems - "this PR didn't create, and didn't make worse" - am I right?   If so, then I suggest - as I did with similar problems raised on this PR - that we deal with it later.  In the spirit of incremental improvement.

Reviewer 4: I think the algorithm does make it worse by breaking a hyphenated word into two words. If the algorithm simply removed the hyphen (ignored it) rather than replacing it with a space, it would raise less false positives. I'm not saying it would be ideal, but less problematic.

Author: To address this I removed this Failed Example, in commit https://github.com/act-rules/act-rules.github.io/commit/1baf1eec0c888d8caf45ebb62ccd217e8039ab41

```html
<a href="#" aria-label="non-standard">nonstandard</a>
```

#### Failed Example 6

The rule has no special handling for acronyms or initialisms.

```html
<a aria-label="WCAG" href="#">W C A G</a>
```

> #### Failed Example 7

Reviewer 4: > The rule has no special handling for abbreviations.

This seems problematic. For many abbreviations it is much more natural (in English) to say the full word when encountering an abbreviations (for example, street abbreviations in an address like "Rd.", "St.", "Pl.", etc).

Perhaps it is just another shortcoming of this SC, but the calculation seems to me to need to be more fuzzy, and bias towards passing, not failing an abbreviation construct.

> The rule has no special handling for abbreviations.

Reviewer 4: > The rule has no special handling for abbreviations.

This seems problematic. For many abbreviations it is much more natural (in English) to say the full word when encountering an abbreviations (for example, abbreviations in an address like "Rd.", "St.", "Pl.", etc).

Perhaps it is just another shortcoming of this SC, but the calculation seems to me to need to be more fuzzy, and bias towards passing, not failing an abbreviation construct.

Author: Again, I agree, but I consider this problem to be in the category of problems that this PR didn't create and didn't make worse.  What do you think?

Reviewer 4: I guess two points.

The SC wording is: "the [name](https://www.w3.org/WAI/WCAG21/Understanding/label-in-name.html#dfn-name) contains the text that is presented visually."

We already have a rule stating that punctuation can be disregarded, so in this example, the displayed wording can be interpreted as "University Ave" and the name is "University Avenue". The accessible name literally contains the text that is presented visually (plus three more characters). So why is it a fail? It should be a pass, no?

The second point is just about abbreviations as synonyms. I can think of situations in English where it would be unusual to say only the abbreviation. For instance, if someone sees "Main St." or "First Rd.", they are going to say the abbreviation as the full word ("main street", "first road"). On the other hand "Ave" may be spoken as the full word or as the abbreviation. Part of me believes it is reasonable to compile a set of abbreviations which are commonly spoken as the full word, and not the abbreviation. But that seems unsustainable/unscalable.

On consideration, I reluctantly agree that failing a mismatch where the name does not contain all the characters of the displayed abbreviation (less punctuation) makes sense.

Author: For future reference, if nothing else: here are some supporting arguments for failing abbreviations, from WCAG.  These are independent of arguments about user impact.  This is just about the letter of the law.  

1) "the exact string of the visible label does not occur in the accessible name" is mentioned as a reason for failure in failure technique F96 for this SC (https://www.w3.org/WAI/WCAG22/Techniques/failures/F96.html).
2) "The accessible name contains a match for the string of the visible label" is a necessary condition for passing, under the "Procedure" section of - again - failure technique F96 for this SC.
3) "When in doubt ... match the string exactly" - "Understanding SC 2.5.3: Label in Name" (https://www.w3.org/WAI/WCAG22/Understanding/label-in-name) 

1 says "exact string" and 3 says "string exactly".  That doesn't allow an abbreviation to pass.

Reviewer 6: While it's a good aspect to be thinking about, I don't agree that this is a problem for this test. From my experience (anecdotal of course) AT users want to have access to the same information as it is presented visually. In what cases would a non-sighted user need to have something fully spelled out and a sighted one wouldn't? While it's less severe of an impact, it's still impactful and should fail this test.

Author: Point taken, but I've since side-stepped this by making anything with abbreviations inapplicable (inspired by per Giacomo's suggestion on Thursday).  I did this in commit https://github.com/act-rules/act-rules.github.io/pull/2075/changes/9490a5e88db20a400adce67408f009c95d40efca .

<button aria-label="Find">search</button>
```

#### Passed Example 7

This button has [visible inner text][] that, according to the [label in name algorithm][], is contained within the [accessible name][]. This example shows why the [label in name algorithm][] uses the [visible inner text][] and not the [visible text content][]: the `<p>` tags insert whitespace into the result in the former but not the latter.

```html
<button aria-label="Hello world">
	<p>Hello</p>
	<p>world</p>
</button>
```

#### Passed Example 8

(Similar to previous example.)  This link has [visible inner text][] that, according to the [label in name algorithm][], is contained within the [accessible name][]. This example shows why the [label in name algorithm][] uses the [visible inner text][] and not the [visible text content][]: the `<h6>` tags insert whitespace into the result in the former but not the latter.

```html
<a href="#" aria-label="Some article by John Doe">
	<h6>Some article</h6>
	<p>by John Doe</p>
</a>
```

#### Passed Example 9

> The [visible inner text][] of this link is "ACT" (with no spaces) because of the explicit styles of `display: inline` on the `p` elements and the absence of whitespace between the `div` elements. The cases of `display: inline` and `display: block` are handled by the definition of [visible inner text of an element][]. This example shows that the definition agrees with the visual rendering done by the browser.

Reviewer: ```suggestion
The [visible inner text][] of this link is "ACT" (with no spaces) because of the explicit styles of `display: inline` on the `div` elements and the absence of whitespace between them. The cases of `display: inline` and `display: block` are handled by the definition of [visible inner text of an element][]. This example shows that the definition agrees with the visual rendering done by the browser.
```

```html
<a href="#" aria-label="ACT">
	<div style="display: inline">A</div><div style="display: inline">C</div><div style="display: inline">T</div>
</a>
```

#### Passed Example 10

The [visible inner text][] is "Download specification". The words "the" and "gizmo" aren't part of it.

```html
<a aria-label="Download specification" href="#"
	>Download <span style="visibility: hidden">the</span> <span style="display: none">gizmo</span> specification</a
>
```

#### Passed Example 11

The [visible inner text][] is "Download specification", which includes a space character between the two words due to the [second clause of the definition of visible inner text of a text node][].

```html
<a aria-label="Download specification" href="#"
	><span>Download</span><span id="space"> </span><span>specification</span></a
>
```

#### Passed Example 12

> This example shows that the [visible inner text][] isn't always the same as the [`innerText` IDL attribute](https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute). The visible inner text is "Download specification". The `innerText` is `Download \ngizmo\nspecification`. This rule uses the visible inner text - not innerText.

Reviewer: ```suggestion
This example shows that the [visible inner text][] isn't always the same as the [`innerText` IDL attribute](https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute). The visible inner text is "Download specification". The `innerText` is `Download \ngizmo\nspecification`. This rule uses the visible inner text — not innerText.
```

> This example shows that the [visible inner text][] isn't always the same as the [`innerText` IDL attribute](https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute). The visible inner text is "Download specification". The `innerText` is `Download \ngizmo\nspecification`. This rule uses the visible inner text - not innerText.

Reviewer: ```suggestion
This example shows that the [visible inner text][] isn't always the same as the [`innerText` IDL attribute](https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute). The visible inner text is "Download specification". The `innerText` is `Download \ngizmo\nspecification`. This rule uses the visible inner text — not innerText.
```

```html
<style>
	.visually-hidden {

> 		/* Source: https://www.tpgi.com/the-anatomy-of-visually-hidden/ */

Reviewer: **Issue:** I am a bit reluctant to link to a vendor's page, as it could be seen as endorsement. This looks like generic techniques and not something they invented, so maybe we can do without the comment?
```suggestion
```

> 		/* Source: https://www.tpgi.com/the-anatomy-of-visually-hidden/ */

Reviewer: **Issue:** I am a bit reluctant to link to a vendor's site, as it could be seen as endorsement. Given that these are fairly standard techniques, maybe we can remove the comment?

```suggestion
```

		clip-path: inset(50%);
		height: 1px;
		overflow: hidden;
		position: absolute;
		white-space: nowrap;
		width: 1px;
	}
</style>
<a aria-label="Download specification" href="#">Download <span class="visually-hidden">gizmo</span> specification</a>
```

#### Passed Example 13

This example shows that the [label in name algorithm][] handles many kinds of whitespace.

```html
<a aria-label="compose email" href="#">
	compose &nbsp;&nbsp;<br />
	email
</a>
```

#### Passed Example 14

This example passes the rule because "YYYY-MM-DD" is in round brackets. Text in round brackets is removed by the [label in name algorithm][], because it is not normally spoken by speech-input users.

```html
<button aria-label="Search by date">Search by date (YYYY-MM-DD)</button>
```

#### Passed Example 15

> This passes for two reasons: 1) because the ellipsis ("…") is [non-text content][], and 2) because the ellipsis is neither a letter nor a digit and so is filtered out by the [label in name algorithm][].

Reviewer: ```suggestion
The ellipsis ("…") is neither a letter nor a digit and so is filtered out by the [label in name algorithm][]; this is correct because it is [non-text content][].
```

Author: My version wasn't great, but it's not clear to me that your version gets us all the way there either.  The line of the algorithm we're referring to is:

> For each character that either a) represents non-text content, or b) isn't a letter or a digit: replace that character with a space character.

So your wording "is neither a letter nor a digit ... this is correct because it is non-text content" bundles together (a) and (b).  Right?  If I'm right, I think it's misleading.

Reviewer: What I was trying to say is in the line of "(technical): ellipsis is not in the Letter or Number unicode classes; (philosophical): and this is a good thing for us because ellipsis matches WCAG definition of 'non-text content' and is ignore by the SC". So "(b), which is indeed the correct behaviour for WCAG" (not really referencing (a) which is not needed in that case).

That would be different from "X for close" which is ignored under (a) only.

```html
<a href="#" aria-label="Discover Italy">Discover It</a>
```

#### Failed Example 4

This link's [accessible name][] contains two tokens (according to the[label in name algorithm][]) and the [visible inner text][] contains one token. So it fails the rule.

```html
<a aria-label="just ice" href="#">justice</a>
```

#### Failed Example 5

The rule has no special handling for acronyms or initialisms.

```html
<a aria-label="WCAG" href="#">W C A G</a>
```

#### Failed Example 6

This link has [visible inner text][] with mathematical symbols and is not contained within the [accessible name][] because the mathematical symbols are represented as English words (not digits) in the accessible name. This is [explicitly mentioned in WCAG](https://www.w3.org/WAI/WCAG22/Understanding/label-in-name#mathematical-expressions-and-formulae).

```html
<a href="/" aria-label="Proof of two multiplied by two is four">Proof of 2&times;2=4</a>
```

#### Failed Example 7

This rule has no special handling for converting mathematical symbols into words, or vice versa.

```html
<button aria-label="11 times 3 equals 33">11×3=33</button>
```

#### Failed Example 8

> This button's accessible name contains the same tokens that are in the visible label. But they aren't in the same order, so it fails the sublist check part of the [label in name algorithm][], and so it fails the rule.

Reviewer: ```suggestion
This button's accessible name contains the same tokens that are in the visible label. But they aren't in the same order, so it fails the contiguous subsequence check part of the [label in name algorithm][], and so it fails the rule.
```
Note: it could/should be an internal link to that point in the definition.

Author: I did this in 08c2a5b260b305ff19fe5a6c7925307194f4d14c.

```html

> <button aria-label="how are you"><span>you</span><span>how</span><span>are</span></button>

Reviewer: **Question:** I am not sure what the inner `span` add to this example. Wouldn't it be the same with just:
```suggestion
<button aria-label="how are you">you how are</button>
```

Author: I think the spans makes each word into a separate text node, and that would wrongly pass the current published version of the rule.  This example was invented by Kasper, here: https://github.com/act-rules/act-rules.github.io/issues/1458

```

#### Failed Example 9

> This button's accessible name contains the word "the" in the middle of it, which causes the sublist check of the [label in name algorithm][] (in particular: the "consecutive" requirement of that check) to fail. So it fails the rule.

Reviewer: ```suggestion
This button's accessible name contains the word "the" in the middle of it, which causes the contiguous subsequence check of the [label in name algorithm][] (in particular: the "contiguous" requirement) to fail. So it fails the rule.
```

Author: I did this in 08c2a5b260b305ff19fe5a6c7925307194f4d14c and 1765efa2339d785fde4007afd8b4c2b7bdc90131.

```html
<button aria-label="Download the specification">Download specification</button>
```

#### Failed Example 10

This link's accessible name contains the same digits that are in the visible label, and in the same order. But they have different spaces and punctuation between them, so they are considered separate tokens. So this fails the rule.

```html

> <a aria-label="1 2 3. 4 5 6. 7 8 9 0" href="tel:1234567890">123.456.7890</a>

Reviewer: ```suggestion
<a aria-label="1 2 3. 5 5 5. 0 1 2 3" href="tel:1235550123">123.555.0123</a>
```
If I read correctly, "555.01XX" is reserved for fictional US numbers.

```

#### Failed Example 11

This rule has no special handling which tries to guess when numbers have the same semantic meaning. It operates on tokens only.

```html
<a href="#2021" aria-label="20 21">2021</a>
```

#### Failed Example 12

This rule has no special handling which tries to guess when numbers have the same semantic meaning. It operates on tokens only.

```html
<a href="#" aria-label="fibonacci: 0 1 1 2 3 5 8 13 21 34">fibonacci: 0112358132134</a>
```

#### Failed Example 13

This rule has no special handling for converting digits into words, or vice versa.

```html
<a href="#2021" aria-label="twenty twenty-one">two thousand twenty-one</a>
```

#### Failed Example 14

This rule has no special handling for converting digits into words, or vice versa.

```html
<a aria-label="two zero two three" href="#">2 0 2 3</a>
```

#### Failed Example 15

> The [label in name algorithm][] works on full words.  That is: it requires that each full word in the visible label ("1" in this case) is equal to a full word in the accessible name ("1a" in this case).  Those two words - "1" and "1a" - are not equal, so this element fails the rule.

Reviewer: ```suggestion
The [label in name algorithm][] works on full words.  That is: it requires that each full word in the visible label ("1" in this case) is equal to a full word in the accessible name ("1a" in this case). Those two words — "1" and "1a" — are not equal, so this link fails the rule.
```

```html
<a aria-label="1a" href="#">1</a>
```

#### Failed Example 16

The definition of [visible inner text][] doesn't treat text any differently if it's excluded from the accessibility tree with aria-hidden. So this rule effectively ignores aria-hidden. So this element fails the rule.

```html
<a aria-label="Download specification" href="#">Download <span aria-hidden="true">gizmo</span> specification</a>
```

### Inapplicable

#### Inapplicable Example 1

This `nav` is not a widget, so the [visible inner text][] does not need to match the [accessible name][].

```html
<nav aria-label="main nav">W3C navigation</nav>
```

#### Inapplicable Example 2

> This email text field does not need to have its [visible inner text][] match the [accessible name][]. The content of a text field shows its value instead of its label; it does not [support name from content][supports name from content]. The label is usually adjacent to the text field instead.

Reviewer: ```suggestion
This email text field has a role of `textbox` which does not [support name from content][supports name from content]; so it does not need to have its [visible inner text][] match the [accessible name][]. The content of a text field shows its value instead of its label. The label is usually adjacent to the text field instead.
```

<button aria-label="Find">search</button>
```

#### Passed Example 7

> This button has [visible inner text][] that, according to the [label in name algorithm][], is contained within the [accessible name][]. This example shows why the [label in name algorithm][] uses the [visible inner text][] and not the [visible text content][]: the <p> tags insert whitespace into the result in the former but not the latter.

Reviewer: ```suggestion
This button has [visible inner text][] that, according to the [label in name algorithm][], is contained within the [accessible name][]. This example shows why the [label in name algorithm][] uses the [visible inner text][] and not the [visible text content][]: the `<p>` tags insert whitespace into the result in the former but not the latter.
```

```html
<button aria-label="Hello world">
	<p>Hello</p>
	<p>world</p>
</button>
```

#### Passed Example 8

> Similar to the previous example, showing how the [label in name algorithm][] handles whitespace.

Reviewer: **Note:** "Similar to previous example" is a risky thing to say in an ACT rule where examples tend to move or be removed 😅 Would be better to duplicate (a bit) the description to avoid that reference breaking in the next rewrite of the rule in a year or two.

Author: An interesting challenge: how to make this both "good for the future" and "good for the present".   I just made an attempt in commit c01ccfa59adb11838b7810a2b476ffaa0b75543c.  Let me know what you think.

```html
<a href="#" aria-label="Some article by John Doe">
	<h6>Some article</h6>
	<p>by John Doe</p>
</a>
```

#### Passed Example 9

> The [visible inner text][] of this link is "ACT" (with no spaces) because of the explicit styles of `display: inline` on the `p` elements and the absence of whitespace between the `p` elements. The cases of `display: inline` and `display: block` are handled by the definition of [visible inner text of an element][]. This example shows that the definition agrees with the visual rendering done by the browser.

Reviewer: **Note:** This is also a case where visible inner text differ from `innerText`, due to step 8 of [the later](https://html.spec.whatwg.org/multipage/dom.html#rendered-text-collection-steps) adding line break count on `p` elements whatever their `display` is 🤔 Maybe using `div` instead would make it simpler?

Author: Nice catch.  I just did it in commit 5aa55e62d11ee88f139cbc6bc5ac02d7e16d3ce6.

Interesting thing I noticed along the way: the whitespace between the `<p>` elements (which are now `<div>` elements) was introduced accidentally AFAICS in commit 56177552ff5aac02e26a7b5c9a9a66b38c6bd91e, which was a merge.  That commit looks like it had some "pretty-printing gone wrong".  It changed the visual rendering of this example to have spaces between the letters - so this example's statement "The [visible inner text][] of this link is "ACT" (with no spaces)" was incorrect.  Anyway, I just removed the whitespace between the elements, so now that statement is correct again.

```html
<a href="#" aria-label="ACT">
	<p style="display: inline">A</p>
	<p style="display: inline">C</p>
	<p style="display: inline">T</p>
</a>
```

#### Passed Example 10

The [visible inner text][] is "Download specification". The words "the" and "gizmo" aren't part of it.

```html
<a aria-label="Download specification" href="#"
	>Download <span style="visibility: hidden">the</span> <span style="display: none">gizmo</span> specification</a
>
```

#### Passed Example 11

> The [visible inner text][] is "Download specification", which includes a space character between the two words due to the second clause of the definition of [visible inner text of a text node][].

Reviewer: **Note:** We can get links inside definitions (i.e. to "the second clause") by explicitly adding a `<dfn id="…">` element (or a `span`, …) there. Convention is that such internal ids start with the definition (or rule) key and a colon (`id="visible-inner-text:foo"`) to ensure deduplication.

See, e.g. [Text Inheriting its Programmatic Language from an Element](https://github.com/act-rules/act-rules.github.io/blob/develop/pages/glossary/text-inheriting-language.md?plain=1#L12) for a definition called by [HTML element language subtag matches language](https://github.com/act-rules/act-rules.github.io/blob/develop/_rules/element-lang-matches-default-language-off6ek.md?plain=1#L324) (who also have its own local anchors in the Applicability).

Author: I just pushed some commits that do this, I think.

```html
<a aria-label="Download specification" href="#"
	><span>Download</span><span id="space"> </span><span>specification</span></a
>
```

#### Passed Example 12

> This example shows that the [visible inner text][] isn't always the same as the [innerText][https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute]. The visible inner text is "Download specification". The innerText is 'Download \ngizmo\nspecification'. This rule uses the visible inner text - not innerText.

Reviewer: ```suggestion
This example shows that the [visible inner text][] isn't always the same as the [`innerText` IDL attribute](https://html.spec.whatwg.org/multipage/dom.html#the-innertext-idl-attribute). The visible inner text is "Download specification". The `innerText` is `Download \ngizmo\nspecification`. This rule uses the visible inner text - not innerText.
```

**Question:** Are there really newlines in the `innerText` given that the `span` element still has a `display: inline` style?

Author: To my surprise: yes.  I think it's the "position: absolute;" in the visually-hidden CSS class that adds the "\n"s.

```html
<a href="#" aria-label="Discover Italy">Discover It</a>
```

#### Failed Example 4

This link's [accessible name][] contains two tokens (according to the[label in name algorithm][]) and the [visible inner text][] contains one token. So it fails the rule.

```html
<a aria-label="just ice" href="#">justice</a>
```

#### Failed Example 5

The rule has no special handling for acronyms or initialisms.

```html
<a aria-label="WCAG" href="#">W C A G</a>
```

#### Failed Example 6

This link has [visible inner text][] with mathematical symbols and is not contained within the [accessible name][] because the mathematical symbols are represented as English words (not digits) in the accessible name. This is [explicitly mentioned in WCAG](https://www.w3.org/WAI/WCAG22/Understanding/label-in-name#mathematical-expressions-and-formulae).

```html
<a href="/" aria-label="Proof of two multiplied by two is four">Proof of 2&times;2=4</a>
```

#### Failed Example 7

> Similar to the previous example. This rule has no special handling for converting mathematical symbols into words, or vice versa.

Reviewer: **Comment:** Same remark about "previous example" being fragile in ACT rules.

```html
<button aria-label="how are you"><span>you</span><span>how</span><span>are</span></button>
```

#### Failed Example 9

This button's accessible name contains the word "the" in the middle of it, which causes the sublist check of the [label in name algorithm][] (in particular: the "consecutive" requirement of that check) to fail. So it fails the rule.

```html
<button aria-label="Download the specification">Download specification</button>
```

#### Failed Example 10

This link's accessible name contains the same digits that are in the visible label, and in the same order. But they have different spaces and punctuation between them, so they are considered separate tokens. So this fails the rule.

```html
<a aria-label="1 2 3. 4 5 6. 7 8 9 0" href="tel:1234567890">123.456.7890</a>
```

#### Failed Example 11

This rule has no special handling which tries to guess when numbers have the same semantic meaning. It operates on tokens only.

```html
<a href="#2021" aria-label="20 21">2021</a>
```

#### Failed Example 12

> Similar to the previous example. This rule has no special handling which tries to guess when numbers have the same semantic meaning. It operates on tokens only.

Reviewer: **Comment:** Another "previous example" reference.

```html

> <a aria-label="fibonacci: 0 1 1 2 3 5 8 13 21 34">fibonacci: 0112358132134</a>

Reviewer: **Issue:** The `a` element needs an `href` in order to be a widget.

```

#### Failed Example 13

This rule has no special handling for converting digits into words, or vice versa.

```html
<a href="#2021" aria-label="twenty twenty-one">two thousand twenty-one</a>
```

#### Failed Example 14

> (Same as above.) This rule has no special handling for converting digits into words, or vice versa.

Reviewer: **Comment:** Another "previous example" reference.

#### Inapplicable Example 3

This `div` element does not have a widget role, so the [visible inner text][] does not need to match the [accessible name][].

```html
<div role="tooltip" aria-label="OK">Next</div>
```

#### Inapplicable Example 4

> This link has no [visible inner text][].

Reviewer: ```suggestion
This link has no [visible text content][].
```
The current Applicability is still referring "visible text content".

```html
<a href="https://w3.org" aria-label="W3C homepage">
	<img src="/test-assets/shared/w3c-logo.png" alt="w3c logo" />
</a>
```

#### Inapplicable Example 5

This link's label contains an abbreviation, so it is not applicable.

```html
<a aria-label="University Avenue" href="#">University Ave.</a>
```

#### Inapplicable Example 6

> This word - non-standard / nonstandard - appears in both the element's accessible name and its visible label, using different hyphenation.  So it's not applicable.

Reviewer: ```suggestion
This word — non-standard / nonstandard — appears in both the element's accessible name and its visible label, using different hyphenation. So it's not applicable.
```
